feat(tests): intercept created executions through queue proxy & kill …

brian-mulier-p opened the pull request:
…them if running after test

brian-mulier-p committed:
feat(tests): intercept created executions through queue proxy & kill them if running after test

loicmathieu commented on tests/src/main/java/io/kestra/core/runners/TestRunner.java:
```diff
@@ -25,6 +29,8 @@
 @Slf4j
 @Singleton
 public class TestRunner implements Runnable, AutoCloseable {
+    public static final ThreadLocal<List<Execution>> testExecutions = ThreadLocal.withInitial(ArrayList::new);
```
Why don't you move this thread local inside the TestQueueFactory?
This would avoid setting it so you would directly access it via the factory

brian-mulier-p replied:
Because the execution emit can come from everywhere including non-test threads (scheduler for eg.) so it will not be shared and retrievable from the test thread which is what we need

loicmathieu commented on tests/src/main/java/io/kestra/core/junit/extensions/KestraTestExtension.java:
```diff
@@ -62,5 +73,49 @@ public void afterTestExecution(ExtensionContext context) throws Exception {
+        try {
+            TestRunner.testExecutions.get().stream()
+                .flatMap(launchedExecution -> executionRepository.findById(launchedExecution.getTenantId(), launchedExecution.getId()).stream())
+                .filter(inRepository -> inRepository.getState().isRunning())
```
```suggestion
                .filter(inRepository -> !inRepository.getState().isTerminated())
```

better to also kill paused/queued exec

loicmathieu commented on tests/src/main/java/io/kestra/core/junit/extensions/KestraTestExtension.java:
```diff
@@ -62,5 +73,49 @@ public void afterTestExecution(ExtensionContext context) throws Exception {
+            .orElseThrow()
+            .getAnnotation(KestraTest.class);
+        if (kestraTest.startRunner()
+            && applicationContext.containsBean(ExecutionRepositoryInterface.class)
```
It's easier to check if there is something in the thread local, in this case you know you can kill exec

loicmathieu commented on tests/src/main/java/io/kestra/core/queues/TestQueueFactory.java:
```diff
@@ -0,0 +1,44 @@
+    @Named(QueueFactoryInterface.EXECUTION_NAMED)
+    @Bean(preDestroy = "close")
+    public QueueInterface<Execution> execution() {
+        return (QueueInterface<Execution>) Proxy.newProxyInstance(this.getClass().getClassLoader(), new Class[]{QueueInterface.class}, (proxy, method, args) -> {
```
I'm not a fan of dynamic proxy usage, not easy to understand such code.
But I guess it's easier than maintaining a delegate....

brian-mulier-p replied:
I know I went for the delegate way at first and then thought about the future us that would have to maintain adding simple methods...

loicmathieu commented on tests/src/main/java/io/kestra/core/junit/extensions/KestraTestExtension.java:
```diff
@@ -62,5 +73,49 @@ public void afterTestExecution(ExtensionContext context) throws Exception {
+                        log.warn("Couldn't kill execution {} after test execution", inRepository.getId(), e);
+                    }
+                });
+        } catch (ConcurrentModificationException e) {
```
I'm not sure I understand why you want to retry and what would cause concurrent modifications here

brian-mulier-p replied:
Because the test that just finished could still be populating executions which would, from the proxy, modify concurrently the list of executions. This is good for us as we want  to detect that. My retry will avoid failing by retrying till a gap between execution addition is available. 10 retries max so no way to be stuck forever and I log it so we will no there is something wrong and in which test :)

loicmathieu reviewed (changes requested):
Overall I would prefer we understand why things are not ended properly but as it seems overly complicated why not going into that direction...

brian-mulier-p committed:
fix(tests): add TestQueueFactory setup in beforeEach

Otherwise the rebuildContext was creating a new applicationContext that doesn't contain the testExecutions

brian-mulier-p commented:
I replied and fix all your comments + I spotted one last issue in context handling that is now fixed.

loicmathieu reviewed (approved):
Still not convinced by what you said about the thread local, if you're not in the test thread, you would not have access to it anyway.

I approved it, but it seems it causes more harm than good as a lot of tests are now failing

loicmathieu commented:
This kind of PR should only be merged if it proved to brings benefits: for ex when launching multiple time you see none of very few flaky tests

brian-mulier-p commented:
It just needs some love & polish, trust the process :+1:

brian-mulier-p committed:
fix(tests): remove ThreadLocal + add distinct on TestExecution cleaner

brian-mulier-p committed:
fix(tests): wait for KILL in TestExecution cleaner

brian-mulier-p commented:
In the end @loicmathieu  I removed the need for ThreadLocal so it's easier to understand + I added a wait after sending the killing message to be 100% sure it doesn't overflow across tests :crossed_fingers:

brian-mulier-p committed:
Merge branch 'develop' into fix/flaky-fix-attempt

brian-mulier-p commented:
Not needed for now
